### ISSUE TITLE
Ensure bailed, found path does not match ensure type (pages/app)

### ISSUE THREAD

**Issue by sbfkcel** [vercel/next.js#53837]
### Verify canary release

- [X] I verified that the issue exists in the latest Next.js canary release

### Provide environment information

```bash
Operating System:
      Platform: darwin
      Arch: arm64
      Version: Darwin Kernel Version 22.5.0: Thu Jun  8 22:21:34 PDT 2023; root:xnu-8796.121.3~7/RELEASE_ARM64_T8112
    Binaries:
      Node: 18.15.0
      npm: 9.5.0
      Yarn: 1.22.19
      pnpm: 7.30.0
    Relevant Packages:
      next: 13.4.14-canary.1
      eslint-config-next: 12.1.5
      react: 18.2.0
      react-dom: 18.2.0
      typescript: 5.1.6
    Next.js Config:
      output: N/A
```

### Which area(s) of Next.js are affected? (leave empty if unsure)

CLI (create-next-app)

### Link to the code that reproduces this issue or a replay of the bug

https://codesandbox.io/p/sandbox/lucid-roentgen-mc8x5s?file=%2Fapp%2Fpage.tsx%3A8%2C1

### To Reproduce

- Customize an interface under pages/api
- The interface is repeatedly requested on the page, and there is a certain probability of wrong capture

### Describe the Bug

<img width="1440" alt="image" src="https://github.com/vercel/next.js/assets/5469785/3ef1fb4d-ceaf-4592-a87f-e8bceadc8753">

### Expected Behavior

Errors should not be thrown, let alone accidentally

<sub>[NEXT-1603](https://linear.app/vercel/issue/NEXT-1603/ensure-bailed-found-path-does-not-match-ensure-type-pagesapp)</sub>

**Comment by juanamd** [comment]
Experiencing the same issue with Next 13.4.13 in dev mode.
Haven't tried on a release build, but I believe this started happening when I upgraded from 13.3

**Comment by CesarBenavides777** [comment]
Also experiencing this!

**Comment by sbfkcel** [comment]
I compared the codes one by one, and it was fine to fall back to 13.4.12.

**Comment by sbfkcel** [comment]
Link to code from this question

https://codesandbox.io/p/sandbox/lucid-roentgen-mc8x5s?file=%2Fapp%2Fpage.tsx%3A8%2C1

**Comment by Lukasdotcom** [comment]
I checked where this issue first appeared in the canary releases and it showed up in 13.4.13-canary.0

**Comment by wearon** [comment]
Just encountered it  next@^13.4.6

`Error: Ensure bailed, found path does not match ensure type (pages/app)
[0]     at ensurePageImpl (node_modules/.pnpm/next@13.4.13_@babel+core@7.22.10_react-dom@18.2.0_react@18.2.0/node_modules/next/dist/server/dev/on-demand-entry-handler.js:483:23)
[0] - error Error: Ensure bailed, found path does not match ensure type (pages/app)
[0]     at ensurePageImpl (node_modules/.pnpm/next@13.4.13_@babel+core@7.22.10_react-dom@18.2.0_react@18.2.0/node_modules/next/dist/server/dev/on-demand-entry-handler.js:483:23) {
[0]   digest: undefined
[0] }`

**Comment by u1x30** [comment]
Upgraded to 13.4.15 and continue to get this issue: 

"err\":{\"name\":\"Error\",\"source\":\"server\",\"message\":\"Ensure bailed, found path does not match ensure type (pages/app)\",\"stack\":\"Error: Ensure bailed, found path does not match ensure type (pages/app)\\n    at ensurePageImpl (/home/user/Documents/software/workspace/project/node_modules/next/dist/server/dev/on-demand-entry-handler.js:483:23)\"

The issue happens when rendering pages from the pages directory and stops application requests.  If can please provide which version to revert to that does not have this error?  Tried 13.4.12 and continues to happen.

Any idea on why error happens and possible work around until it gets resolved?

**Comment by u1x30** [comment]
Thank you for working on this, can you please provide a NextJS version we can use while we wait for the fix to be available? This issue is causing requests to fail which is impacting development progress. Any insight on what version of NextJS to revert to would be greatly appreciated.

Thank you.

**Comment by sbfkcel** [comment]
I submitted a PR to fix that. https://github.com/vercel/next.js/pull/54041

**Comment by niemal** [comment]
> Downgraded to 13.4.11 and it got solved
Yeah thank you, this fixed it for me too.

**Comment by WilliamIPark** [comment]
We lost an entire day to this, pls fix

**Comment by azivkovi** [comment]
13.2.4 fixed it for me. As mentioned in [#54041](https://github.com/vercel/next.js/pull/54041#issuecomment-1685673225), delete your package-lock.json, before installing packages.

**Comment by tc-mccarthy** [comment]
I just attempted to upgrade from 13.4.12 to 13.4.19 via `yarn upgrade` and encountered this issue. I then tried deleting my yarn.lock file and running `yarn install` as there was a comment above about trashing your package-lock.json file but this also proved fruitless. I am not using the `/app` directory in my install -- my current implementation strictly uses `/pages`

**Comment by inderrr** [comment]
Facing the same issue. Tried on v13.4.12 and 13.4.19

![Screenshot-2023-08-27-211415 - Copy](https://github.com/vercel/next.js/assets/85822513/8e67c94a-fd54-47f5-bfe6-1b4bf4a8ca45)

**Comment by charleskoehl** [comment]
Same issue here. Started after I upgraded from 13.4.12 to 13.4.16. Also happens with 13.4.19. My app requests an api route 14 times using async.map (parallel requests) and gets 1-5 of these errors each time. I just downgraded to 13.4.12 and the problem went away. No need to delete yarn.lock in my case.

**Comment by azivkovi** [comment]
> 

That is very interesting. I still have the error @13.4.2. After deleting package-lock and installing clean.

**Comment by azivkovi** [comment]
> I just downgraded to 13.4.12 and the problem went away. No need to delete yarn.lock in my case.

That is very interesting. I still have the error @13.4.2. After deleting package-lock and installing clean..

**Comment by realmers** [comment]
I am using pages directory and have noticed the problem when making a change in a component that is being used by a page file, but not when making a change directly to the page file.

**Comment by BenElferink** [comment]
To fix this issue, disable the `app` dir in your project (assuming you are using `src` dir).

Edit your `next.config.js`:
```js
/** @type {import('next').NextConfig} */

const nextConfig = {
  experimental: {
    appDir: false,
  },
}

module.exports = nextConfig
```

**Comment by realmers** [comment]
That did not solve the issue for me, i am using next.js 13.4.19

**Comment by M4TH76** [comment]
Same here "next": "13.4.19"

**Comment by bsplosion** [comment]
Adding my name to the list of those affected on `13.4.19` which see that it's fixed by reverting down to `13.4.12`.  I'm also exclusively using the `pages` directory, no `app` migration performed yet.

@timneutkens is the provided reproduction sufficient?  This is a frequent enough issue that I've now thought there were client code issues multiple times when it was this bug rearing its head.

**Comment by timneutkens** [comment]
This is being worked on already. @wyattjoh is wrapping up a few test failures around the changes to fix it.

**Comment by cosmictoby** [comment]
Can confirm both 13.4.12 and canary of 13.4.20 has fixed this for me.

**Comment by BintangAldian17** [comment]
> 13.4.12

it's work for me, thanks

**Comment by JimCarnicelli** [comment]
I found one cause of this error. I created an empty folder mirroring something from /pages to /app. I hadn't gotten around to adding the relevant content to it. Deleting that empty folder under /app fixed it.

To reproduce this, try creating /pages/api/something.ts and also creating an /app/api/something folder with no route.ts in it. Create some page that consumes that API call. Refresh your browser a few times. Chances are good the API call will fail with this error. Delete that empty "something" folder. Problem solved.
